framework/DataObjects/DataSet.py

```diff
@@ -1513,37 +1517,59 @@ def _getRealizationFromCollectorByIndex(self,index):
       rlz[var] = vals
     return rlz
 
-  def _getRealizationFromCollectorByValue(self,toMatch,tol=1e-15):
+  def _getRealizationFromCollectorByValue(self, toMatch, noMatch, tol=1e-14):
     """
       Obtains a realization from the collector storage matching the provided index
       @ In, toMatch, dict, elements to match
+      @ In, noMatch, dict, elements to AVOID matching (should not match within tolerance)
       @ In, tol, float, optional, tolerance to which match should be made
       @ Out, r, int, index where match was found OR size of data if not found
       @ Out, rlz, dict, realization as {var:value} OR None if not found
     """
+    if toMatch is None:
+      toMatch = {}
+    if noMatch is None:
+      noMatch = {}
     assert(self._collector is not None)
     # TODO KD Tree for faster values -> still want in collector?
     # TODO slow double loop
-    lookingFor = list(toMatch.values())
-    for r,row in enumerate(self._collector[:,tuple(self._orderedVars.index(var) for var in toMatch.keys())]):
+    matchVars, matchVals = zip(*toMatch.items()) if toMatch else ([], [])
+    avoidVars, avoidVals = zip(*noMatch.items()) if noMatch else ([], [])
+    matchIndices = tuple(self._orderedVars.index(var) for var in matchVars)
+    for r, row in enumerate(self._collector[:]):
       match = True
-      for e,element in enumerate(row):
-        # check for matching based on if a number or not
-        if mathUtils.isAFloatOrInt(element):
-          match &= mathUtils.compareFloats(lookingFor[e],element,tol=tol)
-        else:
-          match &= lookingFor[e] == element
-        # if this element doesn't match, the row doesn't match
-        if not match:
-          break
-      # if each element matched, we found a match, so stop looking
+      # find matches first
+      if toMatch:
+        possibleMatch = self._collector[r, matchIndices]
+        for e, element in enumerate(np.atleast_1d(possibleMatch)):
+          # var = matchVars[e]
+          if mathUtils.isAFloatOrInt(element):
+            match &= mathUtils.compareFloats(matchVals[e], element, tol=tol)
+          else:
+            match &= matchVals[e] == element
+          if not match:
+            break
+      # avoid antimatches if we match so far
+      if match and noMatch:
+        # NOTE there may be multiple entries per var in noMatch
+        possibleAvoid = self._collector[r, tuple(self._orderedVars.index(var) for var in avoidVars)]
+        for e, element in enumerate(np.atleast_1d(possibleAvoid)):
+          var = avoidVars[e]
+          if mathUtils.isAFloatOrInt(element):
+            for avoid in np.atleast_1d(avoidVals[e]):
+              match &= not mathUtils.compareFloats(avoid, element, tol=tol)
+              if not match:
+                break
+          else:
+            match &= element not in np.atleast_1d(avoidVals[e]) # TODO histories?
```

Comment:
I don't know how I would want to TODO it in the future, so I tagged the question here as a starting place to look for a person that is debugging or implementing the feature.

Reply:
We should discuss this during the meeting. I remember we also have TODOs in DataObjects. It seems to me we can add TODO in the method docstring.

Reply:
I think it would be asking a lot to move all the TODO notes in RAVEN into method docstrings, and probably not where they belong. These are intended as developer notes to future development activity, not documented feature requests or issues (we have issue tracking for that).